be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+and allocate rec outside the function
+*/
+static void be_dtm0_log_rec_fini(struct m0_dtm0_log_rec **rec,
+				 struct m0_be_tx         *tx)
+{
+	struct m0_dtm0_log_rec *lrec = *rec;
+
+	M0_ASSERT_INFO(M0_IS0(&lrec->dlr_dtx), "DTX should have been finalised "
+		       "already in m0_dtx0_done().");
+	m0_buf_free(&lrec->dlr_payload);
+	m0_dtm0_tx_desc_fini(&lrec->dlr_txd);
+	m0_free(lrec);
+	*rec = NULL;
+}
+
+/* TODO: change convention of this function to:
+void m0_be_dtm0_plog_rec_fini(struct m0_dtm0_log_rec *rec, ..*tx)
+and allocate rec outside the function
+*/
+static void be_dtm0_plog_rec_fini(struct m0_dtm0_log_rec **dl_lrec,
+				  struct m0_be_dtm0_log   *log,
+				  struct m0_be_tx         *tx)
+{
+	struct m0_be_seg       *seg = log->dl_seg;
+	struct m0_dtm0_log_rec *rec = *dl_lrec;
+
+	M0_BE_FREE_PTR_SYNC(rec->dlr_txd.dtd_ps.dtp_pa, seg, tx);
+	M0_BE_FREE_PTR_SYNC(rec->dlr_payload.b_addr, seg, tx);
+	M0_BE_FREE_PTR_SYNC(rec, seg, tx);
+}
+
+static int m0_be_dtm0_log__insert(struct m0_be_dtm0_log  *log,
+				  struct m0_be_tx        *tx,
+				  struct m0_dtm0_tx_desc *txd,
+				  struct m0_buf          *payload)
+{
+	int                      rc;
+	struct m0_dtm0_log_rec	*rec;
+	struct m0_be_seg	*seg = log->dl_seg;
+
+	if (log->dl_is_persistent) {
+		rc = be_dtm0_plog_rec_init(&rec, tx, seg, txd, payload);
+		if (rc != 0)
+			return rc;
+		lrec_be_tlink_create(rec, tx);
+		lrec_be_list_add_tail(log->dl_list, tx, rec);
+	} else {
+		rc = be_dtm0_log_rec_init(&rec, tx, txd, payload);
+		if (rc != 0)
+			return rc;
+		lrec_tlink_init_at_tail(rec, log->dl_tlist);
+	}
+
+	return rc;
+}
+
+static int be_dtm0_log__set(struct m0_be_dtm0_log	 *log,
+			    struct m0_be_tx		 *tx,
+			    const struct m0_dtm0_tx_desc *txd,
+			    struct m0_buf		 *payload,
+			    struct m0_dtm0_log_rec	 *rec)
+{
+	bool		  is_persistent = log->dl_is_persistent;
+	struct m0_buf    *lpayload         = &rec->dlr_payload;
+	struct m0_be_seg *seg	           = log->dl_seg;
+
+	M0_PRE(m0_dtm0_log_rec__invariant(rec));
+
+	/* Attach payload to log if it is not attached */
+	if (!m0_buf_is_set(lpayload) && m0_buf_is_set(payload)) {
+		if (is_persistent) {
+			lpayload->b_nob = payload->b_nob;
+			M0_BE_ALLOC_BUF_SYNC(lpayload, seg, tx);
+			m0_buf_memcpy(lpayload, payload);
+			M0_BE_TX_CAPTURE_BUF(seg, tx, lpayload);
+			M0_BE_TX_CAPTURE_PTR(seg, tx, lpayload);
+		} else {
+			m0_buf_copy(lpayload, payload);
+		}
+	}
+
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, txd);
+	M0_POST(m0_dtm0_log_rec__invariant(rec));
+	if (is_persistent) {
+		M0_BE_TX_CAPTURE_ARR(seg, tx, rec->dlr_txd.dtd_ps.dtp_pa,
+					      rec->dlr_txd.dtd_ps.dtp_nr);
+	}
+
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_update(struct m0_be_dtm0_log  *log,
+                                      struct m0_be_tx        *tx,
+                                      struct m0_dtm0_tx_desc *txd,
+                                      struct m0_buf          *payload)
```

Comment:
Can these two arguments be "const"?

Reply:
Only payload can be 'const'. I don't think we can make any of the other fields as 'const'.